sdk/nodejs/x/automation/errors.ts

```diff
@@ -0,0 +1,56 @@
+// Copyright 2016-2020, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { CommandResult } from "./cmd";
+
+/**
+ * CommandError is an error resulting from invocation of a Pulumi Command.
+ * This is an opaque error that provides utility functions to detect specific error cases
+ * such as concurrent stack updates (`isConcurrentUpdateError`). If you'd like to detect an additional
+ * error case that isn't currently covered, please file an issue: https://github.com/pulumi/pulumi/issues/new
+ * @alpha
+ */
+export class CommandError extends Error {
+    /** @internal */
+    constructor(private commandResult: CommandResult) {
+        super(commandResult.toString());
+        this.name = "PulumiCommandError";
+    }
+    /**
+     * Returns true if the error was a result of a conflicting update locking the stack.
+     *
+     * @returns a boolean indicating if this failure was due to a conflicting concurrent update on the stack.
+     */
+    isConcurrentUpdateError() {
+        return this.commandResult.stderr.indexOf("[409] Conflict: Another update is currently in progress.") >= 0;
+    }
```

Comment:
Here's an example of usage: https://github.com/pulumi/pulumi/blob/ea9c2c661c4d9b501e60676a196d6a0b0066c62d/sdk/nodejs/x/automation/stack.ts#L95-L102

I'm not sure the proposed approach would work. The underlying implementation of all of these will be `CommandError`. We are parsing command stdout, stderr, and error codes to identify error cases. I don't see how this would work using `instanceof`.

Reply:
For context, the goal of the error design is twofold:

1. Allow users to detect particular failure cases (creating a stack that already exists, an updating failing because it doesn't have the lock).
2. Give us the flexibility to harden this over time. Users shouldn't take dependencies on error code or stderr. We'll maintain this for the time being until we have time to go into the CLI and assign unique error codes to the cases we want to support.

Reply:
The alternative approach that I considered was making these all static methods on `CommandError` so that users don't have to repeat the instanceOf check. Usage at that point would look like: 

```ts
catch(err){
if (CommandError.isConcurrentUpdateError(err)) {
...
}
}
```

Reply:
> I'm not sure the proposed approach would work. The underlying implementation of all of these will be `CommandError`. We are parsing command stdout, stderr, and error codes to identify error cases. I don't see how this would work using `instanceof`.

I was suggesting changing the implementation to enable this, something like:

```typescript
export class ConcurrentUpdateError extends CommandError {
     constructor(commandResult: CommandResult) {
         super(commandResult);
         this.name = "ConcurrentUpdateError";
     }
}

export class StackNotFoundError extends CommandError {
     constructor(commandResult: CommandResult) {
         super(commandResult);
         this.name = "StackNotFoundError";
     }
}

export class StackAlreadyExistsError extends CommandError {
     constructor(commandResult: CommandResult) {
         super(commandResult);
         this.name = "StackAlreadyExistsError";
     }
}

const notFoundRegex = new RegExp("no stack named.*found");
const alreadyExistsRegex = new RegExp("stack.*already exists");
const conflictText = "[409] Conflict: Another update is currently in progress.";

function createCommandError(result: CommandResult): CommandError {
    const stderr = result.stderr;
    return (
        notFoundRegex.test(stderr) ? new StackNotFoundError(result) :
        alreadyExistsRegex.test(stderr) ? new StackAlreadyExistsError(result) :
        stderr.indexOf(conflictText) >= 0 ? new ConcurrentUpdateError(result) :
        new CommandError(result)
    );
}
```

and using the `createCommandError` factory utility function in the implementation when creating the error.

Which allows more idiomatic usage in JS/TS (IMO): https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/try...catch#Conditional_catch-blocks, rather than `err.isConcurrentUpdateError()` or `CommandError.isConcurrentUpdateError(err)` which doesn't feel as idiomatic JS/TS to me.

Reply:
Ah, yeah that's a great idea. I've updated this accordingly.